### PR TITLE
Update philosophy to include other preprocessing classes

### PR DESCRIPTION
**Pull request by stevhliu**
This PR removes the emphasis on NLP and focuses more on `transformers` being designed for all modalities.

### CODE REVIEW

File: docs/source/en/philosophy.mdx [hunk 1]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
 
 🤗 Transformers is an opinionated library built for:
 
-- NLP researchers and educators seeking to use/study/extend large-scale transformers models
-- hands-on practitioners who want to fine-tune those models and/or serve them in production
-- engineers who just want to download a pretrained model and use it to solve a given NLP task.
+- machine learning researchers and educators seeking to use, study or extend large-scale transformers models.
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:17]:
```suggestion
- machine learning researchers and educators seeking to use, study or extend large-scale Transformer models.
```
I always capitalize Transformer :)

File: docs/source/en/philosophy.mdx [hunk 2]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
...
+- hands-on practitioners who want to fine-tune those models and or serve them in production.
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:18]:
```suggestion
- hands-on practitioners who want to fine-tune those models and/or serve them in production.
```

File: docs/source/en/philosophy.mdx [hunk 3]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
...
 The library was designed with two strong goals in mind:
 
 - Be as easy and fast to use as possible:
 
   - We strongly limited the number of user-facing abstractions to learn, in fact, there are almost no abstractions,
     just three standard classes required to use each model: [configuration](main_classes/configuration),
-    [models](main_classes/model) and [tokenizer](main_classes/tokenizer).
+    [models](main_classes/model) and a preprocessing class ([tokenizer](main_classes/tokenizer) for NLP, [feature extractor](main_classes/feature_extractor) for vision and audio, and [processor](main_classes/processors) for multimodal inputs).
   - All of these classes can be initialized in a simple and unified way from pretrained instances by using a common
     `from_pretrained()` instantiation method which will take care of downloading (if needed), caching and
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:29]:
```suggestion
  - All of these classes can be initialized in a simple and unified way from pretrained instances by using a common
    `from_pretrained()` method which will take care of downloading (if needed), caching and
```

File: docs/source/en/philosophy.mdx [hunk 4]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
...
     loading the related class instance and associated data (configurations' hyper-parameters, tokenizers' vocabulary,
     and models' weights) from a pretrained checkpoint provided on [Hugging Face Hub](https://huggingface.co/models) or your own saved checkpoint.
   - On top of those three base classes, the library provides two APIs: [`pipeline`] for quickly
-    using a model (plus its associated tokenizer and configuration) on a given task and
-    [`Trainer`]/`Keras.fit` to quickly train or fine-tune a given model.
+    using a model (plus its associated preprocessing class and configuration) on a given task and
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:33]:
```suggestion
    using a model for inference on a given task and
```

File: docs/source/en/philosophy.mdx [hunk 5]
```diff
@@ -61,20 +61,19 @@ A few other goals:
...
 
 All these classes can be instantiated from pretrained instances and saved locally using two methods:
 
-- `from_pretrained()` lets you instantiate a model/configuration/tokenizer from a pretrained version either
+- `from_pretrained()` lets you instantiate a model, configuration, and preprocessing class from a pretrained version either
   provided by the library itself (the supported models can be found on the [Model Hub](https://huggingface.co/models)) or
-  stored locally (or on a server) by the user,
-- `save_pretrained()` lets you save a model/configuration/tokenizer locally so that it can be reloaded using
+  stored locally (or on a server) by the user.
+- `save_pretrained()` lets you save a model, configuration, and preprocessing class locally so that it can be reloaded using
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:77]: I'd add a third method here which has become increasingly important, namely `push_to_hub`

File: docs/source/en/philosophy.mdx [hunk 6]
```diff
@@ -61,20 +61,19 @@ A few other goals:
 
 The library is built around three types of classes for each model:
 
-- **Model classes** such as [`BertModel`], which are 30+ PyTorch models ([torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module)) or Keras models ([tf.keras.Model](https://www.tensorflow.org/api_docs/python/tf/keras/Model)) that work with the pretrained weights provided in the
+- **Model classes** can be PyTorch models ([torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module)) or Keras models ([tf.keras.Model](https://www.tensorflow.org/api_docs/python/tf/keras/Model)) that work with the pretrained weights provided in the
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:64]: I only see PyTorch and Keras mentioned here, but we also support JAX/FLAX

File: docs/source/en/philosophy.mdx [hunk 7]
```diff
@@ -61,20 +61,19 @@ A few other goals:
...
   library.
-- **Configuration classes** such as [`BertConfig`], which store all the parameters required to build
+- **Configuration classes** store all the parameters required to build
   a model. You don't always need to instantiate these yourself. In particular, if you are using a pretrained model
   without any modification, creating the model will automatically take care of instantiating the configuration (which
   is part of the model).
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:69]:
```suggestion
- **Configuration classes** store the hyperparameters required to build
  a model (such as the number of layers, hidden size). You don't always need to instantiate these yourself. In particular, if you are using a pretrained model
  without any modification, creating the model will automatically take care of instantiating the configuration (which
  is part of the model).
```

File: docs/source/en/philosophy.mdx [hunk 8]
```diff
@@ -61,20 +61,19 @@ A few other goals:
...
-- **Tokenizer classes** such as [`BertTokenizer`], which store the vocabulary for each model and
-  provide methods for encoding/decoding strings in a list of token embeddings indices to be fed to a model.
+- **Preprocessing classes** convert the raw data into a format accepted by the model. A [tokenizer](main_classes/tokenizer) stores the vocabulary for each model and provide methods for encoding and decoding strings in a list of token embeddings indices to be fed to a model. [Feature extractors](main_classes/feature_extractor) preprocess audio or vision inputs, and a [processor](main_classes/processors) handles multimodal inputs.
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:70]:
```suggestion
- **Preprocessing classes** convert the raw data into a format accepted by the model. A [tokenizer](main_classes/tokenizer) stores the vocabulary for each model and provide methods for encoding and decoding strings in a list of token embedding indices to be fed to a model. [Feature extractors](main_classes/feature_extractor) preprocess audio or vision inputs, and a [processor](main_classes/processors) handles multimodal inputs.
```

File: docs/source/en/philosophy.mdx [hunk 9]
```diff
@@ -48,11 +48,11 @@ A few other goals:
...
+  - The preprocessing classes and base model APIs are standardized to easily switch between models.
 
-- Incorporate a subjective selection of promising tools for fine-tuning/investigating these models:
+- Incorporate a subjective selection of promising tools for fine-tuning and investigating these models:
 
-  - A simple/consistent way to add new tokens to the vocabulary and embeddings for fine-tuning.
+  - A simple and consistent way to add new tokens to the vocabulary and embeddings for fine-tuning.
   - Simple ways to mask and prune transformer heads.
 
 - Switch easily between PyTorch and TensorFlow 2.0, allowing training using one framework and inference using another.
```
Comment (NielsRogge) [inline docs/source/en/philosophy.mdx:58]: I'd also add FLAX here

**Review by NielsRogge — APPROVED** [review]
Thanks for improving this! What can perhaps also be added is our "one model, one file" philosophy, where we duplicate code rather than using inheritance to make sure models stay independent (also related to @patrickvonplaten's blog post). 

This is pretty different from other libraries, and it's one that is quite core to the design of the library.

Edit: I see you already refer to it. It's an open question whether our coding philosophy should be included in `philosophy.mdx`, no strong opinion here.

File: docs/source/en/philosophy.mdx [hunk 10]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
...
+    [models](main_classes/model) and a preprocessing class ([tokenizer](main_classes/tokenizer) for NLP, [feature extractor](main_classes/feature_extractor) for vision and audio, and [processor](main_classes/processors) for multimodal inputs).
   - All of these classes can be initialized in a simple and unified way from pretrained instances by using a common
     `from_pretrained()` instantiation method which will take care of downloading (if needed), caching and
     loading the related class instance and associated data (configurations' hyper-parameters, tokenizers' vocabulary,
     and models' weights) from a pretrained checkpoint provided on [Hugging Face Hub](https://huggingface.co/models) or your own saved checkpoint.
   - On top of those three base classes, the library provides two APIs: [`pipeline`] for quickly
-    using a model (plus its associated tokenizer and configuration) on a given task and
-    [`Trainer`]/`Keras.fit` to quickly train or fine-tune a given model.
+    using a model (plus its associated preprocessing class and configuration) on a given task and
+    [`Trainer`] to quickly train or fine-tune a given model.
```
Comment (sgugger) [inline docs/source/en/philosophy.mdx:34]:
```suggestion
    [`Trainer`] to quickly train or fine-tune a PyTorch model (TensorFlow model are all compatible with `Keras.fit`).
```

File: docs/source/en/philosophy.mdx [hunk 11]
```diff
@@ -14,27 +14,27 @@ specific language governing permissions and limitations under the License.
...
   - As a consequence, this library is NOT a modular toolbox of building blocks for neural nets. If you want to
-    extend/build-upon the library, just use regular Python/PyTorch/TensorFlow/Keras modules and inherit from the base
-    classes of the library to reuse functionalities like model loading/saving.
+    extend or build upon the library, just use regular Python, PyTorch, TensorFlow, Keras modules and inherit from the base
+    classes of the library to reuse functionalities like model loading and saving. If you'd like to learn more about our coding philosophy, check out our [Repeat Yourself](https://huggingface.co/blog/transformers-design-philosophy) blog post.
```
Comment (sgugger) [inline docs/source/en/philosophy.mdx:37]:
```suggestion
    classes of the library to reuse functionalities like model loading and saving. If you'd like to learn more about our coding philosophy for models, check out our [Repeat Yourself](https://huggingface.co/blog/transformers-design-philosophy) blog post.
```
This is not what we apply everywhere, only model code files.

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

**Pull request merged**